File: 09_mar/LL_09_mar/helpers/SList.java
package helpers;

import static java.lang.System.out;

public class SList {
	Node head;

	public SList() {
		this.head = null;
	}

	public void addFront(int val) {
		Node tempNode = new Node(val);
		tempNode.next = this.head;
		this.head = tempNode;
	}

	public void addBack(int val) {
		if (this.head == null) {
			this.addFront(val);
			return;
		}

		Node tempNode = new Node(val);
		Node currNode = this.head;

		while (currNode.next != null) {
			currNode = currNode.next;
		}

		currNode.next = tempNode;
	}

	public void printList() {
		Node currNode = this.head;
		String arr = "[ ";

		while (currNode != null) {
			arr += currNode.value + ", ";
			currNode = currNode.next;
		}

		arr += "]";
		out.println(arr);
	}

	public Node back() {
		Node currNode = this.head;

		while (currNode.next != null) {
			currNode = currNode.next;
		}

		return currNode;
	}

	public void removeBack() {
		if (this.head == null) {
			out.println("List is empty");
			return;
		} else if (this.head.next == null) {
			this.head = null;
			return;
		}

		Node currNode = this.head.next,
			prevNode = this.head;

		while (currNode.next != null) {
			prevNode = currNode;
			currNode = prevNode.next;
		}

		prevNode.next = null;
	}

	public void moveMinToFront() {
		if (this.head == null) {
			out.println("List is empty");
			return;
		}

		Node currNode = this.head,
			prevNode = this.head,
			min = this.head;

		while (currNode.next != null) {
			if (currNode.next.value < min.value) {
				min = currNode.next;
				prevNode = currNode;
			}

			currNode = currNode.next;
		}

		if (min == this.head) return;

		prevNode.next = min.next;
		min.next = this.head;
		this.head = min;
	}

	public void moveMaxToBack() {
		if (this.head == null) {
			out.println("List is empty");
			return;
		}

		Node currNode = this.head,
			prevNode = this.head,
			max = this.head;

		while (currNode.next != null) {
			if (currNode.next.value > max.value) {
				max = currNode.next;
				prevNode = currNode;
			}

			currNode = currNode.next;
		}

		if (currNode == max) return;
		if (max == head) head = max.next;
		else prevNode.next = max.next;

		currNode.next = max;
		max.next = null;
	}
}

File: 09_mar/LL_09_mar/helpers/SList.cpp
#include "SList.h"
#include <iostream>
#include <string>

SList::SList()
{
	head = NULL;
}

void SList::addFront(int val)
{
	node * tempNode = new node;
	tempNode -> value = val;
	tempNode -> next = head;
	head = tempNode;
}

void SList::addBack(int val)
{
	if (!head)
	{
		addFront(val);
		return;
	}

	node * tempNode = new node;
	node * currNode = head;
	tempNode -> value = val;
	tempNode -> next = NULL;

	while (currNode -> next)
	{
		currNode = currNode -> next;
	}

	currNode -> next = tempNode;
}

void SList::printList()
{
	node * currNode = head;
	std::string holder = "[ ";

	while (currNode)
	{
		holder += std::to_string(currNode -> value) + ", ";
		currNode = currNode -> next;
	}

	holder += "]";

	std::cout << holder << std::endl;
}

node * SList::back()
{
	if (!head) return head;

	node * currNode = head;

	while (currNode -> next)
	{
		currNode = currNode -> next;
	}

	return currNode;
}

void SList::removeBack()
{
	if (!head)
	{
		std::cout << "List is empty" << std::endl;
		return;
	}
	else if (!head -> next)
	{
		head = NULL;
		return;
	}

	node * prevNode = head;
	node * currNode = head -> next;

	while (currNode -> next)
	{
		prevNode = currNode;
		currNode = prevNode -> next;
	}

	prevNode -> next = NULL;

	delete currNode;
}

void SList::moveMinToFront()
{
	if (!head)
	{
		std::cout << "List is empty" << std::endl;
		return;
	}

	node * prevMin = head;
	node * min = head;
	node * currNode = head;

	while (currNode -> next)
	{
		if (currNode -> next -> value < min -> value)
		{
			prevMin = currNode;
			min = currNode -> next;
		}

		currNode = currNode -> next;
	}

	if (min == head) return;

	prevMin -> next = min -> next;
	min -> next = head;
	head = min;
}

void SList::moveMaxToBack()
{
	if (!head)
	{
		std::cout << "List is empty" << std::endl;
		return;
	}

	node * prevMax = head;
	node * max = head;
	node * currNode = head;

	while (currNode -> next)
	{
		if (currNode -> next -> value > max -> value)
		{
			prevMax = currNode;
			max = currNode -> next;
		}
		currNode = currNode -> next;
	}

	if (currNode == max) return;
	if (max == head) head = max -> next;
	else prevMax -> next = max -> next;

	currNode -> next = max;
	max -> next = NULL;
}

SList::~SList()
{
	node * p = head;
	node * q = head;

	while (q)
	{
		p = q;
		q = p -> next;

		if (q) delete p;
	}
}
